src/liballoc/allocator.rs

```diff
@@ -0,0 +1,786 @@
+    /// satisfying the original request. This condition implies that
+    /// such an allocation request will never succeed on the given
+    /// allocator, regardless of environment, memory pressure, or
+    /// other contextual condtions.
+    ///
+    /// For example, an allocator that does not support zero-sized
+    /// blocks can return this error variant.
+    Unsupported { details: &'static str },
+}
+
+impl AllocErr {
+    pub fn invalid_input(details: &'static str) -> Self {
+        AllocErr::Unsupported { details: details }
+    }
+    pub fn is_memory_exhausted(&self) -> bool {
+        if let AllocErr::Exhausted { .. } = *self { true } else { false }
+    }
+    pub fn is_request_unsupported(&self) -> bool {
+        if let AllocErr::Unsupported { .. } = *self { true } else { false }
+    }
+}
+
+/// The `CannotReallocInPlace` error is used when `fn realloc_in_place`
+/// was unable to reuse the given memory block for a requested layout.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct CannotReallocInPlace;
+
+/// An implementation of `Allocator` can allocate, reallocate, and
+/// deallocate arbitrary blocks of data described via `Layout`.
+///
+/// Some of the methods require that a layout *fit* a memory block.
+/// What it means for a layout to "fit" a memory block means is that
+/// the following two conditions must hold:
+///
+/// 1. The block's starting address must be aligned to `layout.align()`.
+///
+/// 2. The block's size must fall in the range `[use_min, use_max]`, where:
+///
+///    * `use_min` is `self.usable_size(layout).0`, and
+///
+///    * `use_max` is the capacity that was (or would have been)
+///      returned when (if) the block was allocated via a call to
+///      `alloc_excess` or `realloc_excess`.
+///
+/// Note that:
+///
+///  * the size of the layout most recently used to allocate the block
+///    is guaranteed to be in the range `[use_min, use_max]`, and
+///
+///  * a lower-bound on `use_max` can be safely approximated by a call to
+///    `usable_size`.
+///
+pub unsafe trait Allocator {
+
+    /// Returns a pointer suitable for holding data described by
+    /// `layout`, meeting its size and alignment guarantees.
+    ///
+    /// The returned block of storage may or may not have its contents
+    /// initialized. (Extension subtraits might restrict this
+    /// behavior, e.g. to ensure initialization.)
+    ///
+    /// Returning `Err` indicates that either memory is exhausted or `layout` does
+    /// not meet allocator's size or alignment constraints.
+    ///
+    /// Implementations are encouraged to return `Err` on memory
+    /// exhaustion rather than panicking or aborting, but this is
+    /// not a strict requirement. (Specifically: it is *legal* to use
+    /// this trait to wrap an underlying native allocation library
+    /// that aborts on memory exhaustion.)
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<Address, AllocErr>;
```

Comment:
(still, even if its not UB, it doesn't actually solve our own problem, since the `alloc::heap` implementation of `fn check_size_and_alignment` includes a debug assert that `size != 0`, so we would need to either side step that or include an appropriate check up here...)